### PR TITLE
Rabbitmq exchanges metricset

### PR DESCRIPTION
**Pull request by kvalev**
This pull requests add a new metricset for RabbitMQ exchanges.

Related to #6442

### CODE REVIEW

File: metricbeat/docs/modules/rabbitmq.asciidoc [hunk 1]
```diff
@@ -21,7 +21,7 @@ in <<configuration-metricbeat>>. Here is an example configuration:
 ----
 metricbeat.modules:
 - module: rabbitmq
-  metricsets: ["node", "queue", "connection"]
+  metricsets: ["node", "queue", "connection", "exchange"]
```
Comment (ruflin) [inline metricbeat/docs/modules/rabbitmq.asciidoc]: Reminds me we should sort metricsets alphabetical (I can do that in a follow up PR).

File: metricbeat/module/rabbitmq/exchange/_meta/data.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (ruflin) [inline metricbeat/module/rabbitmq/exchange/_meta/data.json]: Any chance you could create a data.json here? The command you need is as following:
```
go test -tags=integration github.com/elastic/beats/metricbeat/module/rabbitmq/exchange/... -data
```

It assumes you have rabbitmq running on localhost.

File: metricbeat/module/rabbitmq/exchange/data.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		"arguments":   c.Dict("arguments", s.Schema{
+		}),
+		"messages":    c.Dict("message_stats", s.Schema{
+			"publish": s.Object{
+				"count": c.Int("publish"),
+				"details": c.Dict("publish_details", s.Schema{
+					"rate": c.Float("rate"),
+				}),
+			},
+			"publish_in": s.Object{
```
Comment (ruflin) [inline metricbeat/module/rabbitmq/exchange/data.go]: We could make the above `publish.total.count` and this one here `publish.in` and the following `publish.out`. Is it correct the the above is just the total?

File: metricbeat/module/rabbitmq/exchange/exchange_test.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	t.Logf("%s/%s event: %+v", f.Module().Name(), f.Name(), event.StringToPrint())
+
+	assert.EqualValues(t, "exchange.name", event["name"])
+	assert.EqualValues(t, "/", event["vhost"])
+	assert.EqualValues(t, true, event["durable"])
+	assert.EqualValues(t, false, event["auto_delete"])
+	assert.EqualValues(t, false, event["internal"])
+
+	messages := event["messages"].(common.MapStr)
```
Comment (ruflin) [inline metricbeat/module/rabbitmq/exchange/exchange_test.go]: I wonder if for the comparison here you could create a `common.MapStr` with all fields inside with the expected values and then compare it? Would probably make the code a bit more readable and easier to extend.

File: metricbeat/module/rabbitmq/exchange/data.go (hunk 3 above)
Comment (jsoriano) [reply metricbeat/module/rabbitmq/exchange/data.go]: The advantage of keeping it as `publish_in` is that all the groups of metrics under `messages` have the same structure, and they are also kept as a more direct mapping of source metrics.

File: metricbeat/module/rabbitmq/exchange/_meta/data.json (hunk 2 above)
Comment (kvalev) [reply metricbeat/module/rabbitmq/exchange/_meta/data.json]: Will do.

File: metricbeat/module/rabbitmq/exchange/data.go (hunk 3 above)
Comment (ruflin) [reply metricbeat/module/rabbitmq/exchange/data.go]: I'm always in favor or having all the related metrics under on group but it's really hard to have a clear rule here. I'm good with both.

Comment (kvalev) [reply metricbeat/module/rabbitmq/exchange/data.go]: TBH I am not sure what "publish" is supposed to be. The [official documentation](https://rawcdn.githack.com/rabbitmq/rabbitmq-management/rabbitmq_v3_6_11/priv/www/doc/stats.html) says "Count of messages published.", but in my environment it is always 0, although publish_in and publish_out are both 0+.
Regarding the naming - it is up to you as I am not familiar with the metric set naming guidelines.

File: metricbeat/module/rabbitmq/exchange/_meta/data.json (hunk 2 above)
Comment (kvalev) [reply metricbeat/module/rabbitmq/exchange/_meta/data.json]: This command does not generate the full json file (the message stats are missing). Am I missing something?

```
{
    "@timestamp": "2017-10-12T08:05:34.853Z",
    "beat": {
        "hostname": "host.example.com",
        "name": "host.example.com"
    },
    "metricset": {
        "host": "localhost:15672",
        "module": "rabbitmq",
        "name": "exchange",
        "rtt": 115
    },
    "rabbitmq": {
        "exchange": {
            "arguments": {},
            "auto_delete": false,
            "durable": true,
            "internal": false,
            "name": "",
            "type": "direct",
            "vhost": "/"
        }
    }
}
```

**Comment by CodingSpiderFox** [comment]
@kvalev will you find the time to finish this PR in the near future? I need this feature soon

**Comment by ruflin** [comment]
@kvalev I case you can't find the time to finish this PR, lets us know so we can take it over (or perhaps @CodingSpiderFox? ) to get it merged.

File: metricbeat/module/rabbitmq/exchange/exchange_test.go (hunk 4 above)
Comment (kvalev) [reply metricbeat/module/rabbitmq/exchange/exchange_test.go]: I tried doing that, but the Equals comparison fails for some reason, although the maps are identical (albeit in a different order): 
```
	Error:      	Not equal:
	            	expected: common.MapStr{"return_unroutable":common.MapStr{"count":40, "details":common.MapStr{"rate":123.123}}, "redeliver":common.MapStr{"count":30, "details":common.MapStr{"rate":0.123}}, "publish":common.MapStr{"count":123, "details":common.MapStr{"rate":0.1}}, "publish_in":common.MapStr{"count":100, "details":common.MapStr{"rate":0.5}}, "publish_out":common.MapStr{"count":99, "details":common.MapStr{"rate":0.9}}, "ack":common.MapStr{"count":60, "details":common.MapStr{"rate":12.5}}, "deliver_get":common.MapStr{"details":common.MapStr{"rate":43.21}, "count":50}, "confirm":common.MapStr{"count":120, "details":common.MapStr{"rate":98.63}}}
	            	actual  : common.MapStr{"ack":common.MapStr{"count":60, "details":common.MapStr{"rate":12.5}}, "deliver_get":common.MapStr{"count":50, "details":common.MapStr{"rate":43.21}}, "confirm":common.MapStr{"count":120, "details":common.MapStr{"rate":98.63}}, "return_unroutable":common.MapStr{"count":40, "details":common.MapStr{"rate":123.123}}, "redeliver":common.MapStr{"count":30, "details":common.MapStr{"rate":0.123}}, "publish":common.MapStr{"count":123, "details":common.MapStr{"rate":0.1}}, "publish_in":common.MapStr{"count":100, "details":common.MapStr{"rate":0.5}}, "publish_out":common.MapStr{"details":common.MapStr{"rate":0.9}, "count":99}}

```

**Comment by CodingSpiderFox** [comment]
I'm not good doing git stuff. Still learning. How can I do the rebase? Fork this repo, pull the branch kvalev and push it to my repo, rebase it on master and then open another PR?

**Comment by kvalev** [comment]
@ruflin I verified it again against my local RabbitMQ instance and it seems that in case the exchanges were not used recently (probably since last restart), the `messages_stats` section is missing. Even if there were some messages, not all subsection will be returned by the RabbitMQ - only the ones that are not 0. This is also the reason why the data.json generation returns a very basic data skeleton. If I understand it correctly, the data generation uses the Dockerfile to start a RabbitMQ instance and fetch the data from there. However given that the instance was just started it does not report any stats back. If I publish some test messages (using `rabbitmqadmin`), parts of the stats sections are present. Should I change the Dockerfile to add some test data, so that the data.json generation returns something meaningful? I dont know whether this will break something else.
I also made all exchange stat sections optional - I hope this was the right way to do it.

**Comment by jsoriano** [comment]
jenkins, test this please

**Comment by ruflin** [comment]
@kvalev I like the idea of populating some data before the test. It could also be part of the test instead of the docker image, like this it also work if done with a local instance.

**Comment by CodingSpiderFox** [comment]
> @kvalev I like the idea of populating some data before the test. It could also be part of the test instead of the docker image, like this it also work if done with a local instance.

So what is the actual TODO here?

**Comment by jsoriano** [comment]
I have created PR #6955 to resolve pending issues

**Comment by ruflin** [comment]
Closing as https://github.com/elastic/beats/pull/6955 was merged. @kvalev Thanks a lot for your work and kicking this off.

**Pull request closed without merging**